packages/flutter/lib/src/material/slider_theme.dart

```diff
@@ -638,20 +618,75 @@ class PaddleSliderValueIndicatorShape extends SliderComponentShape {
     return _bottomLobeEnd;
   }
 
-  void _drawValueIndicator(Canvas canvas, Offset center, Paint paint, double scale,
-      TextPainter labelPainter, double textScaleFactor) {
+  // Determines the "best" offset to keep the bubble on the screen. The calling
+  // code will bound that with the available movement in the paddle shape.
+  double _getIdealOffset(
+    RenderBox parentBox,
+    double halfWidthNeeded,
+    double scale,
+    Offset center,
+    MediaQueryData mediaQueryData,
+  ) {
+    const double edgeMargin = 4.0;
+    final Rect topLobeRect = new Rect.fromLTWH(
+      -_topLobeRadius - halfWidthNeeded,
+      -_topLobeRadius - _distanceBetweenTopBottomCenters,
+      2.0 * (_topLobeRadius + halfWidthNeeded),
+      2.0 * _topLobeRadius,
+    );
+    final Offset origin = center - parentBox.localToGlobal(Offset.zero);
+    // We can just multiply by scale instead of a transform, since we're scaling
+    // around (0, 0).
+    final Offset topLeft = parentBox.localToGlobal((topLobeRect.topLeft * scale) + origin);
+    final Offset bottomRight = parentBox.localToGlobal((topLobeRect.bottomRight * scale) + origin);
+    final Size screenBounds = mediaQueryData?.size ?? (window.physicalSize / window.devicePixelRatio);
+    double shift = 0.0;
+    if (topLeft.dx < edgeMargin) {
+      shift = edgeMargin - topLeft.dx;
+    }
+    if (bottomRight.dx > screenBounds.width - edgeMargin) {
+      shift = screenBounds.width - bottomRight.dx - edgeMargin;
+    }
+    shift = (scale == 0.0 ? 0.0 : shift / scale);
+    return shift;
+  }
+
+  void _drawValueIndicator(
+    RenderBox parentBox,
+    Canvas canvas,
+    Offset center,
+    Paint paint,
+    double scale,
+    TextPainter labelPainter,
+    MediaQueryData mediaQueryData,
+  ) {
     canvas.save();
     canvas.translate(center.dx, center.dy);
     // The entire value indicator should scale with the text scale factor,
     // to keep it large enough to encompass the label text.
-    canvas.scale(scale * textScaleFactor, scale * textScaleFactor);
-    final double inverseTextScale = 1.0 / textScaleFactor;
+    final double textScaleFactor = mediaQueryData?.textScaleFactor ?? window.textScaleFactor;
```

Comment:
never default to the `window` unless you're also going to subscribe to notifications that the `window` has changed. For example, consider how this would react if the user changed the scale factor wile the app was on the screen, in the case where you had fallen back to the default here. you wouldn't update the slider's paint.

Reply:
OK.  I'll default to 1.0 then (or just require a media query in context).